### ISSUE TITLE
Cannot set property length of [object Object]

### ISSUE THREAD

**Issue by oroce** [feross/buffer#79]
Buffer 3.5.0 breaks in Chrome (`45.0.2454.85 (Official Build) (64-bit)`)

What I'm getting is 
~~~
Uncaught TypeError: Cannot set property length of [object Object] which has only a getter
~~~

It happens when buffer tries to set `.prototype.length` to `undefined`, here: https://github.com/feross/buffer/blob/master/index.js#L372

I discovered this while I was rolling out our app's new version. Later I'm gonna try to put together a small demo.

(I'm using browserify 11.11.0 which uses `buffer@3.5.0`)

**Comment by byronanderson** [comment]
+1, causing me to need to start using npm shrinkwrap to get buffer 3.4.3

**Comment by marioatlp** [comment]
+1

**Comment by feross** [comment]
Can someone share code to reproduce this error? I can't trigger it and the whole test suite passes on every supported browser...

![](https://camo.githubusercontent.com/7e992522c7f303e600d0aa87e18aa867d371ee1b/68747470733a2f2f73617563656c6162732e636f6d2f62726f777365722d6d61747269782f6275666665722e737667)

**Comment by dcousens** [comment]
Assuming this bug can be confirmed, what is the best course of action anyway?

**Comment by feross** [comment]
Depends on the bug. We can either try to fix it, or revert the `instanceof` improvement I shipped in v3.5.0.

**Comment by marioatlp** [comment]
The code i had that was causing this to show up was while using webpack and was unfortunately several libraries deep where it was triggered so i don't have a simple test case I know the file that directly had threw the error while using a buffer was https://github.com/primus/primus/blob/master/transformer.js  likely not very helpful but you might be able to see a use case that would break it which is more obvious to you having written the buffer library

**Comment by nolanlawson** [comment]
@marioatlp Can you provide a stacktrace so we know where transformer was calling Buffer?

**Comment by oroce** [comment]
haha it seems I could reproduce the issue. It's not buffer but `babelify` when it is used as global transform breaks buffer.

Showcase:
* `--transform`: http://purposeindustries.github.io/buffer-bug79/index.html
* `--global-transform`: http://purposeindustries.github.io/buffer-bug79/index-global.html

Code: https://github.com/purposeindustries/buffer-bug79

**Comment by jlbruno** [comment]
Looks like we're getting the same issue. 
Also using webpack and babel together.

**Comment by oroce** [comment]
@jlbruno in webpack: add `node_modules` to the `exclude` list

~~~
module: {
  loaders: [
    {
      test: /\.jsx?$/,
      exclude: /node_modules/,
      loader: 'babel'
    }
  ]
~~~~

**Comment by marioatlp** [comment]
@oroce That solution is only helpful if the issue is a .jsx file in node_modules my issue was in standard JS. an alternative work around until this is fixed is  to add to `package.json`  dependencies section 
  ```  "buffer": "3.4.3", ```
and to add to your webpack config file:

```
plugins: [
		new webpack.ContextReplacementPlugin(/buffer/, require('buffer'))
	]
```

**Comment by jlbruno** [comment]
Ok, thanks....looks like 3 solutions have worked. Shrinkwrap (locking buffer to 3.4.3 by editing node-libs-browser package.json), excluding node_modules, and adding buffer at a top level dependency locked to v 3.4.3. 

Thanks all. Time to figure out which is the least of the evils.

**Comment by foray1010** [comment]
[node problem](https://github.com/nodejs/node/pull/2931), confirm fixed in node `4.1.1`

**Comment by lostrouter** [comment]
found one other solution for webpack based on 
> ps: you can solve the global transform by simply ignoring buffer: browserify index.js --global-transform [ babelify --ignore buffer ] -o dist.js
```
loaders: [
    { test: /\.jsx$/, loader: 'babel?stage=0&ignore=buffer' }
]
```

**Comment by nelix** [comment]
@lostrouter I would recommend not running babel on third party code, as often the transforms can break code that is unaware of es6. Such as making modules option strict by default.

**Comment by Ryuno-Ki** [comment]
I cross-referenced this issue against browserify, too. They should at least be aware of this issue.

Since I use traceur instead of babel, my solution was `traceur --script ajar.js --out out/ajar.es6.js && browserify out/ajar.es6.js --ignore buffer -o out/ajar.js`, whereby `ajar.js` is the main entry point in that project.

**Comment by mikemaccana** [comment]
3.4.3 still seems to have the broken code:

    $ grep version node_modules/buffer/package.json
      "version": "3.4.3"
    $ grep 'length = undefined' node_modules/buffer/index.js
    Buffer.prototype.length = undefined

**Comment by JSteunou** [comment]
Same here using webpack with https://github.com/mistadikay/strict-loader 

```
<exception>: TypeError: Cannot set property length of [object Object] which has only a getter at Object.eval (webpack:///../~/webpack/~/node-libs-browser/~/buffer/index.js?:376:25) at eval (webpack:///../~/webpack/~/node-libs-browser/~/buffer/index.js?:1548:30) at ...
BP: undefined
Buffer: Buffer(arg)
INVALID_BASE64_RE: undefined
MAX_ARGUMENTS_LENGTH: undefined
SlowBuffer: SlowBuffer(subject, encoding)
allocate: allocate(that, length)
asciiSlice: asciiSlice(buf, start, end)
asciiToBytes: asciiToBytes(str)
asciiWrite: asciiWrite(buf, string, offset, length)
base64: Object
base64Slice: base64Slice(buf, start, end)
base64ToBytes: base64ToBytes(str)
base64Write: base64Write(buf, string, offset, length)
base64clean: base64clean(str)
binarySlice: binarySlice(buf, start, end)
binaryWrite: binaryWrite(buf, string, offset, length)
blitBuffer: blitBuffer(src, dst, offset, length)
byteLength: byteLength(string, encoding)
checkIEEE754: checkIEEE754(buf, value, offset, ext, max, min)
checkInt: checkInt(buf, value, offset, ext, max, min)
checkOffset: checkOffset(offset, ext, length)
checked: checked(length)
decodeCodePointsArray: decodeCodePointsArray(codePoints)
fromArray: fromArray(that, array)
fromArrayBuffer: fromArrayBuffer(that, array)
fromArrayLike: fromArrayLike(that, array)
fromBuffer: fromBuffer(that, buffer)
fromJsonObject: fromJsonObject(that, object)
fromNumber: fromNumber(that, length)
fromObject: fromObject(that, object)
fromString: fromString(that, string, encoding)
fromTypedArray: fromTypedArray(that, array)
global: Window
hexSlice: hexSlice(buf, start, end)
hexWrite: hexWrite(buf, string, offset, length)
ieee754: Object
isArray: isArray()
kMaxLength: kMaxLength()
objectWriteUInt16: objectWriteUInt16(buf, value, offset, littleEndian)
objectWriteUInt32: objectWriteUInt32(buf, value, offset, littleEndian)
rootParent: Object
slowToString: slowToString(encoding, start, end)
stringtrim: stringtrim(str)
this: Object
toHex: toHex(n)
typedArraySupport: typedArraySupport()
ucs2Write: ucs2Write(buf, string, offset, length)
utf8Slice: utf8Slice(buf, start, end)
utf8ToBytes: utf8ToBytes(string, units)
utf8Write: utf8Write(buf, string, offset, length)
utf16leSlice: utf16leSlice(buf, start, end)
utf16leToBytes: utf16leToBytes(str, units)
writeDouble: writeDouble(buf, value, offset, littleEndian, noAssert)
writeFloat: writeFloat(buf, value, offset, littleEndian, noAssert)
```

**Comment by JSteunou** [comment]
Ok I ran the test suit inside the same browser I have the bug in, and it is all good. Maybe someone (webpack?) is doing a freeze on the object, or something else.

**Comment by JSteunou** [comment]
Find it, don't add / use strict mode on Buffer.js https://github.com/mistadikay/strict-loader/issues/5#issuecomment-165024914

**Comment by scottburch** [comment]
I did a horrible patch in webpack for this issue that others might be able to use:

----  strict-stripper.js ------
module.exports = source => source.replace(/['"]use strict['"]/, '');

---- in webpack.config.js (loaders section) ------
{test: /buffer/, loader: `${__dirname}/strict-stripper.js`},

**Comment by nolanlawson** [comment]
So I take it the steps to fix this bug are:

1. Add `"use strict" to the top of [index.js](https://github.com/feross/buffer/blob/96f9561e55019fd530e48886366eda74449bf002/index.js)
2. Re-run the tests
3. Fix stuff until it works

In general, it looks like this could be fixed by simply hoisting the `function Buffer () { ... }` declaration to the top of the file.

**Comment by nolanlawson** [comment]
I can confirm that I can reproduce this by just adding `'use strict'` to the top of `index.js` and then running `npm run test-browser-local`. I'm not sure how to fix it.

**Comment by nolanlawson** [comment]
^ fixed!

**Comment by JSteunou** [comment]
nicely done, thank you!

**Comment by feross** [comment]
Nicely done! Fix released as 3.6.0. And just like that, we're down to 0 open issues and 0 open PRs! :+1:
